### PR TITLE
code completion

### PR DESCRIPTION
**Pull request by megaflo**
This pr introduces a first version for code completion via antlr. Column names and tables are collected via a custom ast visitor.

### CODE REVIEW

File: src/antlr/CompletionsVisitor.ts [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+ */
+interface Result {
+  queryInfo: QueryInfo;
+  scope: Scope;
+  snippets: Snippet[];
+}
+
+/**
+ * A visitor for the SQL ast that collects all required information for the auto completion.
+ * The position for which to run completion will be identified via a special token (CARET: '<|>').
```
Comment (fabien0102) [inline src/antlr/CompletionsVisitor.ts:175]: I'm not really fan of changing the grammar to have this special caret... The node range should be enough, so let's not mix everything.

Let's talk on this on Monday.

**Comment by chris-baynes** [comment]
Since quite a bit of the diff is the generated code we could try `linguist-generated=true` on those files to improve readability of the PR, see https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/customizing-how-changed-files-appear-on-github
wdyt @megaflo?

**Review by fabien0102 — CHANGES_REQUESTED** [review]
This code completion logic should not leak inside the grammar

**Comment by megaflo** [comment]
The approach has been changed to avoid grammar changes. Instead of parsing and checking a special token the visitor now identifies the cursor by a special string suffix ("_CURSOR_"). From the grammars/antlrs point of view this is just (part of) a regular identifier. Contrary to what I initially assumed the resulting code is not more complex (or hacky) than the token version, so I am happy with this change.

**Comment by micha-f** [comment]
@megaflo Shall we close this one?

**Comment by megaflo** [comment]
Yep :)

**Pull request closed without merging**